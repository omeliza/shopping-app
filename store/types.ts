export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images: string[];
  quantity?: number;
}

export interface IUIState {
  drawerIsOpen: boolean;
  loginModalIsOpen: boolean;
  registerModalIsOpen: boolean;

  toggleDrawer: () => void;
  toggleLoginModal: () => void;
  toggleRegisterModal: () => void;
}

export interface ICartState {
  cart: IProduct[];
  totalItems: number;
  totalPrice: number;

  addToCart: (item: IProduct) => void;
  removeFromCart: (item: IProduct) => void;
  clearCart: () => void;
}
