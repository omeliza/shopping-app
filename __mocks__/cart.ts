export const cart = [
  {
    id: 1,
    title: 'Product 1',
    description: 'Description of product 1',
    price: 200,
    discountPercentage: 10,
    rating: 4,
    stock: 94,
    brand: 'Brand 1',
    category: 'category 1',
    thumbnail: 'https://picsum.photos/200/300',
    images: [''],
  },
  {
    id: 2,
    title: 'Product 2',
    description: 'Description of product 2',
    price: 300,
    discountPercentage: 20,
    rating: 4,
    stock: 94,
    brand: 'Brand 2',
    category: 'category 2',
    thumbnail: 'https://picsum.photos/200/400',
    images: [''],
  },
];
