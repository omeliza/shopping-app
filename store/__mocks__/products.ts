export const product1 = {
  id: 1,
  title: 'Product 1',
  description: 'Description of product 1',
  price: 200,
  discountPercentage: 10,
  rating: 4,
  stock: 94,
  brand: 'Brand 1',
  category: 'category 1',
  thumbnail: '',
  images: [''],
};
export const product2 = {
  id: 2,
  title: 'Product 2',
  description: 'Description of product 2',
  price: 300,
  discountPercentage: 15,
  rating: 4,
  stock: 20,
  brand: 'Brand 2',
  category: 'category 2',
  thumbnail: '',
  images: [''],
};
