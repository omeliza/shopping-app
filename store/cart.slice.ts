import { StateCreator } from "zustand";
import { ICartState } from "./types";

export const createCartSlice: StateCreator<ICartState> = (set, get) => ({
  cart: [],
  totalItems: 0,
  totalPrice: 0,

  addToCart: (product) => {
    const cart = get().cart;
    const cartItem = cart.find((item) => item.id === product.id);

    if (cartItem) {
      const updatedCart = cart.map((item) =>
        item.id === product.id
          ? { ...item, quantity: (item.quantity as number) + 1 }
          : item
      );
      set((state) => ({
        cart: updatedCart,
        totalItems: state.totalItems + 1,
        totalPrice: state.totalPrice + product.price,
      }));
    } else {
      const updatedCart = [...cart, { ...product, quantity: 1 }];
      set((state) => ({
        cart: updatedCart,
        totalItems: state.totalItems + 1,
        totalPrice: state.totalPrice + product.price,
      }));
    }
  },

  removeFromCart: (product) =>
    set((state) => ({
      cart: state.cart.filter((item) => item.id !== product.id),
      totalItems: state.totalItems - 1,
      totalPrice: state.totalPrice - product.price,
    })),

  clearCart: () =>
    set(() => ({
      cart: [],
      totalItems: 0,
      totalPrice: 0,
    })),
});
