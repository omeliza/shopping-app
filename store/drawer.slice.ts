import { StateCreator } from "zustand";

import { IDrawerState } from "./types";

export const createDrawerSlice: StateCreator<IDrawerState> = (set) => ({
  isOpen: false,
  toggle: () =>
    set((state) => ({
      isOpen: !state.isOpen,
    })),
});
