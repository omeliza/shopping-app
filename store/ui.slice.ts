import { StateCreator } from 'zustand';

import { IUIState } from './types';

export const createUISlice: StateCreator<IUIState> = (set) => ({
  drawerIsOpen: false,
  loginModalIsOpen: false,
  registerModalIsOpen: false,

  toggleDrawer: () =>
    set((state) => ({
      drawerIsOpen: !state.drawerIsOpen,
    })),
  toggleLoginModal: () =>
    set((state) => ({
      loginModalIsOpen: !state.loginModalIsOpen,
    })),
  toggleRegisterModal: () =>
    set((state) => ({
      registerModalIsOpen: !state.registerModalIsOpen,
    })),
});
