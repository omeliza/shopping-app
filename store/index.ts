import { create } from "zustand";

import { ICartState, IDrawerState } from "./types";
import { createDrawerSlice } from "./drawer.slice";
import { createSelectors } from "./selectors";
import { createCartSlice } from "./cart.slice";
import { persist } from "zustand/middleware";

const useBaseBoundStore = create<IDrawerState & ICartState>()(
  persist(
    (...a) => ({
      ...createDrawerSlice(...a),
      ...createCartSlice(...a),
    }),
    {
      name: "storage",
    }
  )
);

export const useBoundStore = createSelectors(useBaseBoundStore);
