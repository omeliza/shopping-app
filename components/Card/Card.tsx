"use client";
import { useBoundStore } from "@/store";
import { IProduct } from "@/store/types";
import Image from "next/image";

interface IProps {
  product: IProduct;
}

export default function Card({ product }: IProps) {
  const { description, price, thumbnail, title } = product;
  const addToCart = useBoundStore.use.addToCart();

  return (
    <div className="flex flex-col bg-white rounded-lg shadow-lg p-4 hover:shadow-xl">
      <Image
        src={thumbnail}
        alt={title}
        width={100}
        height={100}
        className="h-20 w-full object-contain"
      />
      <div className="flex flex-col justify-between flex-1">
        <h2 className="text-lg font-semibold">{title}</h2>
        <p className="text-gray-600 flex-1">{description}</p>
        <div className="flex items-center justify-between mt-4">
          <span className="text-gray-800 font-semibold">{price}$ </span>
          <button
            className="ml-2 bg-blue-500 text-white py-2 px-4 rounded-lg hover:bg-blue-600"
            onClick={() => addToCart(product)}
          >
            Add to Cart
          </button>
        </div>
      </div>
    </div>
  );
}
