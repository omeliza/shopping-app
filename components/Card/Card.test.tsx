import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Card from './Card';
import { useBoundStore } from '@/store';
import { product1 } from '@/__mocks__/products';

jest.mock('@/store');

describe('Card', () => {
  it('should render correctly', () => {
    const { container } = render(<Card product={product1} />);
    expect(container).toMatchSnapshot();
  });

  it('should add product to cart', () => {
    const { getByText } = render(<Card product={product1} />);
    userEvent.click(getByText('Add to Cart'));

    waitFor(() => {
      expect(useBoundStore.use.addToCart).toHaveBeenCalled();
      expect(useBoundStore.use.addToCart).toHaveBeenCalledWith(product1);
    });
  });
});
