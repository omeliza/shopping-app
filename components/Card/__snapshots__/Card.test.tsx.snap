// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly 1`] = `
<div>
  <div
    class="flex flex-col bg-white rounded-lg shadow-lg p-4 hover:shadow-xl"
  >
    <img
      alt="Product 1"
      class="h-20 w-full object-contain"
      data-nimg="1"
      decoding="async"
      height="100"
      src=""
      style="color: transparent;"
      width="100"
    />
    <div
      class="flex flex-col justify-between flex-1"
    >
      <h2
        class="text-lg font-semibold"
      >
        Product 1
      </h2>
      <p
        class="text-gray-600 flex-1"
      >
        Description of product 1
      </p>
      <div
        class="flex items-center justify-between mt-4"
      >
        <span
          class="text-gray-800 font-semibold"
        >
          200
          $ 
        </span>
        <button
          class="ml-2 bg-blue-500 text-white py-2 px-4 rounded-lg hover:bg-blue-600"
        >
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
